toolkit/types/csaf/csaf.go

```diff
@@ -0,0 +1,492 @@
+	return csafDoc.ProductTree.Relationships.FindRelationship(productID, category)
+}
+
+// ProductBranch methods
+
+// FindFirstProduct recursively searches for the first product identifier in the tree
+// and returns it or an empty string if no product identifier is found.
+func (branch *ProductBranch) FindFirstProduct() string {
+	if branch.Product.ID != "" {
+		return branch.Product.ID
+	}
+
+	// No nested branches
+	if branch.Branches == nil {
+		return ""
+	}
+
+	// Recursively search for the first product	identifier
+	for _, b := range branch.Branches {
+		if p := b.FindFirstProduct(); p != "" {
+			return p
+		}
+	}
+
+	return ""
+}
+
+// FindFirstProductName recursively searches for the first product name in the tree
+// and returns it or an empty string if no product name is found.
+func (branch *ProductBranch) FindFirstProductName() string {
+	if branch.Product.Name != "" {
+		return branch.Product.Name
+	}
+
+	// No nested branches
+	if branch.Branches == nil {
+		return ""
+	}
+
+	// Recursively search for the first product	identifier
+	for _, b := range branch.Branches {
+		if p := b.FindFirstProductName(); p != "" {
+			return p
+		}
+	}
+
+	return ""
+}
+
+// FindProductIdentifier recursively searches for the first product identifier in the tree
+// given the helper value.
+func (branch *ProductBranch) FindProductIdentifier(helperType, helperValue string) *Product {
+	if len(branch.Product.IdentificationHelper) != 0 {
+		for k := range branch.Product.IdentificationHelper {
+			if k != helperType {
+				continue
+			}
+			if branch.Product.IdentificationHelper[k] == helperValue {
+				return &branch.Product
+			}
+		}
+	}
+
+	// No nested branches
+	if branch.Branches == nil {
+		return nil
+	}
+
+	// Recursively search for the first identifier
+	for _, b := range branch.Branches {
+		if p := b.FindProductIdentifier(helperType, helperValue); p != nil {
+			return p
+		}
+	}
+
+	return nil
+}
+
+// FindProductIdentifier recursively searches for the first product identifier in the tree
+// given the productID.
+func (branch *ProductBranch) FindProductByID(productID string) *Product {
+	if branch.Product.ID == productID {
+		return &branch.Product
+	}
+
+	// No nested branches
+	if branch.Branches == nil {
+		return nil
+	}
+
+	// Recursively search for the first product id
+	for _, b := range branch.Branches {
+		if p := b.FindProductByID(productID); p != nil {
+			return p
+		}
+	}
+
+	return nil
+}
+
+type ProductList []Product
+
+// ListProducts returns a flat list of all products in the branch
+func (branch *ProductBranch) ListProducts() ProductList {
+	list := ProductList{}
+	list.Add(branch.Product)
+	for _, b := range branch.Branches {
+		for _, p := range b.ListProducts() {
+			list.Add(p)
+		}
+	}
+	return list
+}
+
+// ProductList methods
+
+// Add adds a prodocut to the product list if its not there, matching id and
+// software identifiers.
+func (pl *ProductList) Add(p Product) {
```

Comment:
What happens when I add two `Product`s with `ID == ""`? Doesn't `ID` always have to be set?